gslab_scons/tests/test_build_python.py

```diff
@@ -3,40 +3,124 @@
+        os.remove(log_path)
+
     def test_bad_extension(self):
-        env = ''
+        '''
+        Test that build_python() raises the appropriate error
+        when the first source lacks the appropriate file extension.
+        '''
         with self.assertRaises(BadExtensionError):
-            build_python('../build/py.py', 
-                         ['bad', './input/python_test_script.py'], 
-                         env)
-   
+            gs.build_python('./build/py.py', 
+                           ['bad', './input/python_test_script.py'], 
+                           env = '')
+
+    def test_unintended_inputs(self):
+        '''
+        Test that build_python() handles unintended inputs
+        as we expect it to. 
+        '''
+        env    = None
+        source = './input/python_test_script.py'
+        target = './build/py.py'
+        log    = './build/sconscript.log'
+        # env can be None
+        gs.build_python(target, source, env)
+        self.check_log(log)
+
+        #== target ===============
+        # Containers of strings should not raise errors
+        gs.build_python(('a', 'b'), source, env)           
+        self.check_log('./sconscript.log')
+        # Neither should empty strings
+        gs.build_python('', source, env)           
+        self.check_log('./sconscript.log')       
+        # Other non-string inputs should
+        with self.assertRaises(TypeError):
+            gs.build_python(None, source, env)
+        with self.assertRaises(TypeError):
+            gs.build_python(1, source, env)
+
+        #== source ===============
+        # We can have multiple sources, and we can 
+        # specify sources that don't exist...
+        gs.build_python(target = '', 
+                        source = ['./input/python_test_script.py', 
+                                 'nonexistent_data.txt'], 
+                        env    = None)           
+        self.check_log('./sconscript.log')
+        # ...but a .py script must be the first source.
+        with self.assertRaises(BadExtensionError):
+            gs.build_python(target = '', 
+                            source = ['nonexistent_data.txt',
+                                   './input/python_test_script.py'], 
+                            env    = None)
+        # build_python() does not raise an error if this .py file
+        # doesn't actually exist (this may be undesirable).
+        # Instead, we expect it to print an error message to the 
+        # console and continue. 
+        print "Expecting console output suggesting an error:\n"
+        gs.build_python(target = '', 
+                        source = ['nonexistent_data.py',
+                              './input/python_test_script.py'], 
+                         env    = None)
+        print "No longer expecting an error message.\n"
+        self.check_log('./sconscript.log')
+
+    def test_target_creation_unnecessary(self):
```

Comment:
SCons will notice that a target wasn't created successfully, but this just causes it to re-run the next time to try and build the target. Adding something that checks for targets being created properly could be nice so we don't have more 'silent' target misspecification errors. (If you agree, a new task to implement probably makes sense.)